docs/madr/decisions/055-default-hostname-generators.md

```diff
@@ -0,0 +1,218 @@
+# Default HostnameGenerators
+
+* Status: accepted
+
+## Context and Problem Statement
+
+We introduced HostnameGenerator for MeshService and MeshExternalService which is also compatible with upcoming MeshMultizoneService.
+This MADR proposes default generators taking into account different scenarios.
+
+## Considered Options
+
+Template and matching:
+* Proposed options
+
+Shipping:
+* Document
+* Create
+
+Suffix:
+* `.svc.{{ type }}.{{ zone }}`
+* `.{{ zone }}.{{ type }}`
+
+## Decision Outcome
+
+Template and matching - Proposed options.
+Shipping - short term document. Once we get all prerequisites and MeshService/MeshExternalService/MeshMultizoneService goes GA, we create by default.
+Suffix - ?
+
+## Pros and Cons of the Options - Template and matching
+
+### Prerequisites
+
+**Labels**
+* `kuma.io/origin`. The value is either `zone` or `global`.
+  The initial idea was to preserve the label between Zones, so that `kuma.io/origin: zone` in Zone A means that it's also the same on Global and on Zone B (if resource is synced).
+  However, we diverged from this, so that Zone A has `kuma.io/origin: zone`, Global has `kuma.io/origin: zone` and Zone B has `kuma.io/origin: global` (because it was synced from global).
+  We could adjust the implementation to initial assumptions, but it requires careful handling. If we were to do this we would have to introduce something like `kuma.io/synced: "true"`.
+  We found existing state to be useful for hostname generators, and we did not find good use case to leverage this label preservation across zones.
+* `kuma.io/env`. The value is either `kubernetes` or `universal`.
+  Introducing this label is out of scope of this MADR.
+
+**Display name**
+`{{ .DisplayName }}` resolves to `kuma.io/display-name` if it exists and it fallbacks to name.
+We can remove the fallback if we add `kuma.io/display-name` to Universal resources by default.
+It's better than `{{ label "kuma.io/display-name" }}`, because looking at Kube resources it's annotation, not a label, so it might be confusing for users.
+
+### MeshService
+
+#### Local Kubernetes MeshService
+
+It's rather unlikely for user to use custom hostnames to consume MeshService that is local in the Kubernetes cluster.
+Therefore, no default HostnameGenerator is needed for it.
+
+#### Synced Kubernetes MeshService
+
+```yaml
+spec:
+  template: '{{ label "k8s.kuma.io/service-name" }}.{{ label "k8s.kuma.io/namespace" }}.svc.mesh.{{ label "kuma.io/zone" }}'
```

Comment:
Why don't we want `.namespace`?

Reply:
why do we need namespace? I think it's easier to just rely on existing mechanism, no?

Reply:
Easier for who though? For UX, readability, optimizing for most common use cases I think `.Namespace` is better. Labels are completely arbitrary and so using a label to get a namespace, which is guaranteed to be there for a Kubernetes service, is excessively noisy. It adds friction between Kuma and Kubernetes. IMO if namespace doesn't deserve its own field then nothing deserves its own field.

Reply:
Yeah they would all be fields then IMO.

> But for synced resources Namespace is system namespace. Isn't it more clear that label is actually the original namespace?

I highly suspect the intuition is to think "the namespace" of a synced service as its original namespace, rather than the object with a hashed name in `kuma-system`. This synced object is essentially just Kuma internal state. As evidenced by the unusable name.

Unfortunately, I realize this kind of clashes with the decision on `targetRef.namespace`, which doesn't refer to the original namespace for synced services, and you have to use labels. It's really kind of a different scenario because the targetRef semantics are not under our control. I suppose it's arguably a valid reason to use labels here too. But it'd be unfortunate IMO.

Reply:
We did namespace.